crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,406 @@
+
+    /// Return relevant unspents.
+    pub fn try_list_chain_utxos<'a, C>(
+        &'a self,
+        chain: C,
+    ) -> impl Iterator<Item = Result<FullTxOut<ObservedIn<A>>, C::Error>> + 'a
+    where
+        C: ChainOracle + 'a,
+        ObservedIn<A>: ChainPosition,
+    {
+        self.try_list_chain_txouts(chain)
+            .filter(|r| !matches!(r, Ok(txo) if txo.spent_by.is_none()))
+    }
+
+    pub fn list_chain_utxos<'a, C>(
+        &'a self,
+        chain: C,
+    ) -> impl Iterator<Item = FullTxOut<ObservedIn<A>>> + 'a
+    where
+        C: ChainOracle<Error = Infallible> + 'a,
+        ObservedIn<A>: ChainPosition,
+    {
+        self.try_list_chain_utxos(chain)
+            .map(|r| r.expect("error is infallible"))
+    }
+
+    pub fn try_balance<C, F>(
+        &self,
+        chain: C,
+        tip: u32,
+        mut should_trust: F,
+    ) -> Result<Balance, C::Error>
+    where
+        C: ChainOracle,
+        ObservedIn<A>: ChainPosition + Clone,
+        F: FnMut(&Script) -> bool,
+    {
+        let mut immature = 0;
+        let mut trusted_pending = 0;
+        let mut untrusted_pending = 0;
+        let mut confirmed = 0;
+
+        for res in self.try_list_chain_txouts(&chain) {
+            let txout = res?;
+
+            match &txout.chain_position {
+                ObservedIn::Block(_) => {
+                    if txout.is_on_coinbase {
+                        if txout.is_mature(tip) {
+                            confirmed += txout.txout.value;
+                        } else {
+                            immature += txout.txout.value;
+                        }
+                    }
+                }
+                ObservedIn::Mempool(_) => {
+                    if should_trust(&txout.txout.script_pubkey) {
+                        trusted_pending += txout.txout.value;
+                    } else {
+                        untrusted_pending += txout.txout.value;
+                    }
+                }
+            }
+        }
+
+        Ok(Balance {
+            immature,
+            trusted_pending,
+            untrusted_pending,
+            confirmed,
+        })
+    }
+
+    pub fn balance<C, F>(&self, chain: C, tip: u32, should_trust: F) -> Balance
+    where
+        C: ChainOracle<Error = Infallible>,
+        ObservedIn<A>: ChainPosition + Clone,
+        F: FnMut(&Script) -> bool,
+    {
+        self.try_balance(chain, tip, should_trust)
+            .expect("error is infallible")
+    }
+
+    pub fn try_balance_at<C>(&self, chain: C, height: u32) -> Result<u64, C::Error>
+    where
+        C: ChainOracle,
+        ObservedIn<A>: ChainPosition + Clone,
+    {
+        let mut sum = 0;
+        for txo_res in self.try_list_chain_txouts(chain) {
+            let txo = txo_res?;
+            if txo.is_spendable_at(height) {
+                sum += txo.txout.value;
+            }
+        }
+        Ok(sum)
+    }
+
+    pub fn balance_at<C>(&self, chain: C, height: u32) -> u64
```

Comment:
I think I didn't do this because `Balance` has additional fields for "unconfirmed" balance which I couldn't make sense of for this function.

Reply:
> I think I didn't do this because Balance has additional fields for "unconfirmed" balance which I couldn't make sense of for this function.

I meant an `Amount`! Sorry

Reply:
Reference: #823 

I don't think this should be tackled in this PR though.